### PR TITLE
fix(typescript-estree): pass extraFileExtensions to projectService

### LINKED ISSUE #8899: Bug: parserOptions.projectService does not respect extraFileExtensions
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

When linting a project containing e.g. svelte files, you need to provide `extraFileExtensions: ['.svelte']` to make the parser use the correct tsconfig for the specific file. This works well with `project: true`, since that path ends up in `createWatchProgram` where we provide the `extraFileExtensions` option to ts.

However, if we switch `EXPERIMENTAL_useProjectService` on, we end up in a code path where the extraFileExtensions are not passed to ts. In my reproduction case this is surfaced by the fact that ts will then use the default compiler options, which transpiles the code back to ES5, instead of using the correct tsconfig.

While this reproduction example is using Svelte, the same issue should apply to other cases where `extraFileExtensions` would be used (such as when using Vite).

### Reproduction Repository Link

https://github.com/alfredringstad/typescript-eslint-svelte-use-project-service-bug-reproduction

### Repro Steps

1) clone the repo
2) `npm i`
3) `npm run lint`

Switch `EXPERIMENTAL_useProjectService` to `false` in `eslint.config.js` and run `npm run lint` to see how it works well without the project service.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `7.6.0` |
| `@typescript-eslint/typescript-estree`            | `7.6.0` |
| `TypeScript`                           | `5.4.5` |
| `ESLint`                               | `8.57.0` |

### PR DESCRIPTION
**Pull request by alfredringstad**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8899 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Uses `setHostConfiguration` to pass `extraFileExtensions` to project service, when using `EXPERIMENTAL_useProjectService`. This fixes an issue where project service would not detect the appropriate tsconfig file when using non-default file extensions.

The tests are checking that the configFilePath is set in the compiler options. If it's not, it's an indication that the default inferred project is used. I duplicated the test cases used for the non-experimental projects option.

### CODE REVIEW

File: packages/typescript-estree/src/parser-options.ts [hunk 1]
```diff
@@ -123,6 +123,11 @@ export interface ProjectServiceOptions {
    * @default 8
    */
   maximumDefaultProjectFileMatchCount_THIS_WILL_SLOW_DOWN_LINTING?: number;
+
+  /**
+   * Additional file extensions which should be considered in the TypeScript Program compilation.
+   */
+  extraFileExtensions?: string[];
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/parser-options.ts:130]: [Docs] Nit: would you mind putting these options in alphabetical order please? It helps keep the page well structured & easy to navigate.

File: packages/typescript-estree/src/parser-options.ts
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/parser-options.ts:1]: [Docs] `docs/packages/Parser.mdx` will need to be updated for this too.

File: packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts:1]: [Testing] What do you think about adding a more direct unit test to this file too? IMO it's nice to have tests that directly check the API boundaries of where we talk to TypeScript too.

File: packages/typescript-estree/tests/lib/parse.test.ts [hunk 2]
```diff
@@ -367,6 +367,22 @@ describe('parseAndGenerateServices', () => {
             alignErrorPath(error as Error);
           }
         };
+      const testParseExperimentalProjectService =
+        (filePath: string, extraFileExtensions: string[] = ['.vue']) =>
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/parse.test.ts:371]: [Testing] A couple of thoughts:

```suggestion
      const testExtraFileExtensions =
        (filePath: string, extraFileExtensions: string[]) =>
```

* This test is basically only used for `extraFileExtensions`, so let's name it like that
* IMO it's nice to be a bit more explicit on what settings are actually being tested

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Source code changes look great to me! And checking that the config file path is found as a form of testing is very snazzy. Nicely done. 😄

Just requesting changes on a few small things that I'm open to being pushed back on - what do you think?

File: packages/typescript-estree/src/parser-options.ts (hunk 1 above)
Comment (alfredringstad) [reply packages/typescript-estree/src/parser-options.ts]: I had another look at this, and realised this must've been a left-over from a previous iteration, this change is actually not needed, since we rely on the extraFileExtensions of the parent object (same as the non-project-service setup).

File: packages/typescript-estree/src/parser-options.ts
Comment (alfredringstad) [reply packages/typescript-estree/src/parser-options.ts:1]: See comment above, the change was reverted.

File: packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts
Comment (alfredringstad) [reply packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts:1]: Good idea! I added two tests to be extra specific, to verify that we only call setHostConfiguration when extraFileExtensions has been provided.

File: packages/typescript-estree/tests/lib/parse.test.ts (hunk 2 above)
Comment (alfredringstad) [reply packages/typescript-estree/tests/lib/parse.test.ts]: Followed your suggestion. Just want to highlight that this is specifically testing extraFileExtensions when projectService is true, and that there are other tests to cover extraFileExtensions when projectService is not used.

**Comment by alfredringstad** [comment]
Thanks for the feedback @JoshuaKGoldberg! I've implemented your suggestions. Seems like CI failure is due to issues in the main v8 branch, the typescript-estree tests seem to work fine.

File: packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/useProgramFromProjectService.test.ts:1]: [Praise] 💯

File: packages/typescript-estree/tests/lib/parse.test.ts [hunk 3]
```diff
@@ -479,6 +494,48 @@ describe('parseAndGenerateServices', () => {
           `);
         });
       });
+
+      describe('"parserOptions.extraFileExtensions" is non-empty and EXPERIMENTAL_useProjectService is true', () => {
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/parse.test.ts:498]: We _just_ applied this rename recently ✨ 

```suggestion
      describe('"parserOptions.extraFileExtensions" is non-empty and projectService is true', () => {
```

File: packages/typescript-estree/tests/lib/parse.test.ts [hunk 4]
```diff
@@ -362,11 +361,27 @@ describe('parseAndGenerateServices', () => {
...
               extraFileExtensions,
               filePath: join(PROJECT_DIR, filePath),
+              project: './tsconfig.json',
             });
           } catch (error) {
             alignErrorPath(error as Error);
           }
         };
+      const testExtraFileExtensions =
+        (filePath: string, extraFileExtensions: string[] = ['.vue']) =>
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/parse.test.ts:371]: [Testing] I think this got missed in the refactors?

```suggestion
        (filePath: string, extraFileExtensions: string[]) =>
```

I can just apply it myself, about to merge.

**Review by JoshuaKGoldberg — APPROVED** [review]
This is great, thanks! :shipit: 

Normally I'd wait for a bit to let someone else review but I would like to unblock folks who need the extra file extensions from using the project.
